aptos-move/framework/aptos-framework/sources/permissioned_signer.move

```diff
@@ -0,0 +1,423 @@
+            let PermStorage::V1 { perms } = move_from<PermStorage>(permission_addr);
+            simple_map::destroy(perms, |_dk| {}, |_dv| {});
+        }
+    }
+
+    inline fun remove_permission_address(master_addr: address, permission_addr: address) acquires GrantedPermissionHandles {
+        if(exists<GrantedPermissionHandles>(master_addr)) {
+           let granted_permissions = borrow_global_mut<GrantedPermissionHandles>(master_addr);
+            let (found, idx) = vector::index_of(&granted_permissions.active_handles, &permission_addr);
+            if(found) {
+                vector::swap_remove(&mut granted_permissions.active_handles, idx);
+            };
+        };
+    }
+
+    public fun signer_from_permissioned(p: &PermissionedHandle): signer {
+        signer_from_permissioned_impl(p.master_addr, p.permission_addr)
+    }
+
+    public fun signer_from_storable_permissioned(p: &StorablePermissionedHandle): signer {
+        assert!(
+            timestamp::now_seconds() < p.expiration_time,
+            error::permission_denied(E_PERMISSION_EXPIRED)
+        );
+        assert!(
+            exists<PermStorage>(p.permission_addr),
+            error::permission_denied(E_PERMISSION_REVOKED)
+        );
+        signer_from_permissioned_impl(p.master_addr, p.permission_addr)
+    }
+
+    public fun revoke_permission_handle(s: &signer, permission_addr: address) acquires GrantedPermissionHandles, PermStorage {
+        assert!(!is_permissioned_signer(s), error::permission_denied(ENOT_MASTER_SIGNER));
+        let master_addr = signer::address_of(s);
+        destroy_permission_address(permission_addr);
+        remove_permission_address(master_addr, permission_addr);
+    }
+
+    public fun revoke_all_handles(s: &signer) acquires GrantedPermissionHandles, PermStorage {
+        assert!(!is_permissioned_signer(s), error::permission_denied(ENOT_MASTER_SIGNER));
+        let master_addr = signer::address_of(s);
+        if(!exists<GrantedPermissionHandles>(master_addr)) {
+            return
+        };
+
+        let granted_permissions = borrow_global_mut<GrantedPermissionHandles>(master_addr);
+        let delete_list = vector::trim_reverse(&mut granted_permissions.active_handles, 0);
+        vector::destroy(delete_list, |address| {
+            destroy_permission_address(address);
+        })
+    }
+
+    public fun permission_address(p: &StorablePermissionedHandle): address {
+        p.permission_addr
+    }
+
+    public fun assert_master_signer(s: &signer) {
+        assert!(!is_permissioned_signer(s), error::permission_denied(ENOT_MASTER_SIGNER));
+    }
+
+    /// =====================================================================================================
+    /// Permission Management
+    ///
+
+    /// Authorizes `permissioned` with the given permission. This requires to have access to the `master`
+    /// signer.
+    public fun authorize<PermKey: copy + drop + store>(
```

Comment:
The reason why I needed copy initially is because smart table need copy for the key. Not sure what would be the generic requirement for your container.

Reply:
oh, keys. in OrderedMap, there is no restriction. in BigOrderedMap keys need to be copy, values don't.